rfd/0201-native-auto-approval.md

```diff
@@ -0,0 +1,369 @@
+
+## Details
+This feature will be supported by a new `access-native` plugin. This native
+plugin will implement the same interface as already existing access plugins.
+The plugin will be hosted by Teleport and it will be enabled by default. Users
+do not configure or enable/disable this plugin.
+
+The native plugin will rely on a similar workflow that supports access request
+notification routing with access monitoring rules. The plugin watches for Access
+Monitoring Rules (AMR) events and Access Requests (AR) events. If an incoming AR
+matches an existing AMR condition, then the plugin will attempt to automatically
+approve the request.
+
+Note: I've outlined two different approaches below, describing how we could
+modify the access monitoring rules to support auto approvals. Approach (A1) is
+more user friendly for configuring auto approval rules. Approach (A2) is less
+user friendly, but it is more easily extensible to support future use cases.
+Approach (A1) is probably the better approach to complete our goals for this
+RFD, but open to exploring other ideas that could be more compatible with both
+goals.
+
+### Access Monitoring Rule (A1)
+The access monitoring rule now supports a configurable `automatic_approval` spec.
+This spec will contain a specified plugin `name` along with optional conditions
+that are specific to each plugin. Separate conditions are required because each
+plugin relies on different conditions to determine if a user is pre-approved for
+the request. For example, the PagerDuty plugin relies on PagerDuty Services to
+determine if a user is on-call, while this native plugin might rely on the
+user's level, team, and location traits.
+
+To support automatic approvals based on specific user traits, the AMR will be
+extended to support these additional fields:
+- **automatic_approval.name**: This field specifies the plugin that this AMR
+applies for handling auto approvals.
+- **automatic_approval.native**: This field contains auto approval conditions
+specific to the native plugin.
+- **automatic_approval.native.level**: This field specifies the requesting
+user's level.
+- **automatic_approval.native.team**: This field specifies the requesting user's
+team.
+- **automatic_approval.native.location**: This field specifies the requesting
+user's location.
+
+```yaml
+# This example AMR matches ARs that request the "cloud-dev" role. The plugin is
+# then responsible for sending notifications to the "#dev-cloud" Slack channel.
+# The plugin then has access to read the automatic_approval spec to check if the
+# requesting user is pre-approved for the request.
+kind: access_monitoring_rule
+version: v1
+metadata:
+  name: cloud-dev-pre-approved
+spec:
+  notification:
+    name: teleport-slack
+    recipients: ["#dev-cloud"]
+  automatic_approval:
+    name: teleport-native
+    native:
+       level: ["L1"]
+       team: ["Cloud"]
+       location: ["Seattle"]
```

Comment:
As mentioned elsewhere, I think this should definitely be more flexible and allow to filter by user traits and roles. See my proposed example below.